### DISCUSSION TITLE
Negative runoff fluxes in CESM3 Development Simulations

### DISCUSSION THREAD

**Discussion by olyson** [ESCOMP/CTSM#1835] [General]
In the CESM co-chairs meeting of 7/19/22, the OMWG noted that there were significant negative runoff fluxes in several of the recent CESM3 development coupled simulations, in particular, around several small islands.
We of course have had negative runoff fluxes in previous versions of the model (e.g., CESM2), due to irrigation and lakes (negative P-E).
However, these particular islands have neither irrigation nor lakes.
After a bit of analysis and discussion (@olyson @swensosc @wwieder ), we found that there are non-zero land fractions (identified by landfrac on the CLM history file) for gridcells where we don't have any surface data (pfts, lakes, glaciers, urban, crop) on the surface dataset being used. These are being modeled as wetlands which can have negative P-E and therefore negative runoff. 
Several of these islands did not exist in CESM2.  
For CESM3, there seems to be a bit of a mismatch between the landfrac that CLM actually uses (generated from the mesh file?) and the land/ocean mask on the surface dataset.
Question: Does this go away when we generate new surface datasets or will there always potentially be some kind of a mismatch?
@wwieder @swensosc @ekluzek @billsacks @dlawrenncar 

I'm putting our solution at the top of this discussion for easier tracking
Three parts to this plan discussed at the SE meeting

- [x] **Changing wetland to bare ground** @billsacks has ideas to review this and suggested doing this at runtime (not in the surface dataset), see #1878, and #1886
- [x] **Runoff direct to outlet**, this would be a mosart namelist change in coupled simulations that @ekluzek added in dev.116
- [x] **Remove remaining negative runoff** from global river discharge.  This seems like it should happen in the river model (or the coupler). The later would be nice because it only has to be done once, but may be the most challenging.   This is still TBD. 
- [x] #2222

**Comment by billsacks** [comment]
This could change somewhat with newer raw data, but this issue will potentially always exist given our current method for generating surface datasets. Basically: if some area is outside the ocean model grid then CTSM becomes the owner of that area; if none of CTSM's raw datasets claim ownership of that area, then it is set as wetland. (Though note that we're currently doing this slightly wrong - see https://github.com/ESCOMP/CTSM/issues/1716 .)

If this negative runoff is a problem we will need to either (1) come up with a different scheme for what to do when CTSM owns an area that appears to be ocean according to CTSM's raw data sets; (2) change the physics in some way to avoid generating negative runoff.

**Reply by wwieder** [reply]
@billsacks where are we on this first action item, _changing wetlands to bare ground_?  Has this been implemented in a place where it can be used for the next round of couple model simulations?

**Reply by billsacks** [reply]
Confirming what we discussed on the call today: the fix for this was applied to the ctsm5.2 branch, so isn't yet on master. However, as I noted in #1890  we could apply these directly to master if that would be helpful for the upcoming coupled testing:

> Note that this PR is into the ctsm5.2.mksurfdata branch. Most of the changes could be applied directly to master, but there are some related changes to mksurfdata that motivated me to make this PR into the ctsm5.2.mksurfdata branch. However, standard f09 surface datasets don't show any changes due to the code changes in mksurfdata here, and the mksurfdata changes aren't critical anyway, so it would be reasonable to rebase the non-mksurfdata changes onto master if we want them available sooner.

**Reply by wwieder** [reply]
Thanks for checking on this.  As Frank noted this won't really impact MOM science, I don't know it's critical to bring this to main sooner than a CTSM5.2 merge.  We'll see if we're told differently at co-chair's tomorrow.

**Comment by wwieder** [comment]
Thanks for looking into this @olyson.  
@billsacks I agree with your suggested solutions, which of these seem like the best option?For option_2 @swensosc suggested we could consider creating _desert_  (bare ground) islands instead of _wetland_ islands, which should avoid negative runoff.

**Comment by billsacks** [comment]
I'm not sure off-hand. For changing what we call these areas (wetland vs. desert, etc.), I'm worried that solving one problem will create another – like treating substantial areas as bare land that should really be water. Changing the physics to avoid negative runoff sounds like a good idea, but I think could pose some real challenges. Regardless of what we do, I feel like it's going to be hard to avoid having some situations where we generate persistently negative (P-E), and it's hard for me to envision a good solution that avoids relying on the ocean to solve the problem.

**Comment by olyson** [comment]
We asked Cecile to run a short fully coupled simulation similar to one of the CESM3-dev simulations (b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.011) to see what the effects are of taking the negative runoff directly from the river mouths.  
This was accomplished by setting bypass_routing_option='direct_to_outlet' and qgwl_runoff_option='negative' in the mosart namelist. 
Four years of this new simulation have been completed (b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.015). 
Results are shown in slides 7 and 8 of this slide deck:

https://docs.google.com/presentation/d/1gEK5ASR3CNbQPkdCCPm6SIavPjX7BCYh/edit?usp=sharing&ouid=104026900087612515425&rtpof=true&sd=true

Slide 7 shows that TOTAL_DISCHARGE_TO_OCEAN_LIQ is now positive everywhere in the new simulation (015) unlike in the control (011).
Slide 8 (monthly minimum of the lrunoff field from MOM) shows that there are very few ocean grid cells that have negative runoff in 015.  Note the scale for 015 is two orders of magnitude smaller than 011.

There are standard diagnostics here:

https://webext.cgd.ucar.edu/B1850/b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.015/lnd/b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.015.2_4-b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.011.2_4/setsIndex.html

In particular, set7 is of interest.  The variability in these coupled simulations is probably swamping the effects of this change, particularly since these are only 3 year averages, but I don't see anything alarming in the results for discharge of individual rivers or into the individual ocean basins.

**Comment by olyson** [comment]
There are BUDGET WARNINGs in the mosart log file.  For tracer 1 which I assume is liquid water.

/glade/scratch/hannay/archive/b.cesm3_cam058_mom_c.B1850WscMOM.f09_L58_t061.015/logs/rof.log.5167742.chadmin1.ib0.cheyenne.ucar.edu.220726-082955.gz

**Comment by olyson** [comment]
Sean has fixed a couple of bugs in the MOSART code with respect to the BUDGET WARNINGs we encountered when we set bypass_routing_option='direct_to_outlet' and qgwl_runoff_option='negative' in the mosart namelist.  
I ran a pair of land-only simulations, one a control (with bypass_routing_option set to the default value of 'direct_in_place') and one with 'direct_to_outlet'.  There are now no warnings in the log files.
Slide 3 of the slide deck shows the effect of this on reducing the negative TOTAL_DISCHARGE_TO_OCEAN_LIQ.  What remains is due to, "For example, for the caspian, those gridcells don't have an outlet on the ocean mask.  And the little dots in the continental interiors I think are 'sink' points, which don't touch the ocean.  So we would still have negative runoff at those locations that is too large to be compensated for by the rivers.  it would get passed to the ocean at those locations just like the direct_in_place method."

https://docs.google.com/presentation/d/1gEK5ASR3CNbQPkdCCPm6SIavPjX7BCYh/edit?usp=sharing&ouid=104026900087612515425&rtpof=true&sd=true

I don't see any discernable impact on our set7 discharge to the ocean diagnostics as shown here:

https://webext.cgd.ucar.edu/I2000/ctsm51_ctsm51d100_1deg_GSWP3V1_directtooutlet_2000/lnd/ctsm51_ctsm51d100_1deg_GSWP3V1_directtooutlet_2000.6_10-ctsm51_ctsm51d100_1deg_GSWP3V1_2000.6_10/set7/set7.html

I'm going to work on quantifying the magnitude of negative runoff with respect to total runoff and the reduction we get using this method.

**Comment by ekluzek** [comment]
Adding @nmizukami to this issue as we will want to be aware of these issues in mizuRoute. We should for example make sure we implement direct_to_outlet in mizuRoute because of this issue. So direct_in_place and direct_to_outlet for bypass_routing_option should both be implemented in mizuRoute. And any fixes we need to put in place in MOSART we'll also want to get into mizuRoute.

**Comment by ekluzek** [comment]
@olyson do you have a branch of MOSART with these changes? We should get these into a PR for MOSART. You could start the PR even before we are done with it's evaluation.

**Comment by billsacks** [comment]
Thanks @olyson and @swensosc . My understanding is that we still want to move ahead with trying the replacement of wetlands with bare ground in addition to this MOSART fix; is that right? If so, I'll work on it today.

**Comment by olyson** [comment]
Yes, that is my understanding as well, thanks.

**Comment by wwieder** [comment]
Yes, that would be great, @billsacks, as Sean's mosart changes don't help with the negative runoff from wetland islands.  

@olyson, I recall you saying last week that frozen runoff in the log files was different with 'direct_to_outlet' option.  Is this something we should investigate?

**Comment by olyson** [comment]
Yes, definitely should understand that.
Another thing is that in the land-only simulations, global negative runoff was reduced by about 70% (from -7.2e6 m3/s to -2.2e6 m3/s), but I don't see that reflected in the global sum of TOTAL_DISCHARGE_TO_OCEAN_LIQ.  There is a difference between the two simulations, but the difference is quite small.  Looking at the spatial map of the difference field there are a significant number of grid cells that have an increase in TOTAL_DISCHARGE_TO_OCEAN_LIQ which are apparently offsetting the decreases I would expect to see.  I guess I would not expect to see any gridcells like that @swensosc ?

**Comment by swensosc** [comment]
Another difference between the simulations is that the qgwl_runoff_option =
"negative" option is used in the new simulation, but qgwl_runoff_option =
"threshold" is used in the default.  I think that means that in the
default, some of the negative runoff reduces the amount of water in the
river if possible, and the remainder is sent to the coupler.  I don't see
how that would cause the result you are seeing though, but perhaps it's
worth doing a direct_in_place simulation with the negative option so we can
compare.

**Comment by olyson** [comment]
Ok, I think this is fine actually.  We would expect the global sum of TOTAL_DISCHARGE_TO_OCEAN_LIQ to be conserved.
I'm running a 'negative' option simulation anyway.

**Comment by olyson** [comment]
I don't see any significant differences with the 'negative' option combined with 'direct_in_place'.

**Comment by billsacks** [comment]
The question I raised in the co-chairs meeting was about the many inland grid cells that showed negative discharge to ocean in the MOSART history files. My interpretation was that discharge to ocean in these inland cells would only arise from negative contributions, because any positive runoff would be routed to the ocean and so would show up in coastal grid cells. The alternative is that there really are a huge number of grid cells with a negative net runoff over the year, which would be very surprising to me.

**Comment by swensosc** [comment]
I think in mosart there are terminal points that are not on the coasts
('sink' points).  I think runoff at these points would be passed to the
ocean at those locations, not the coasts.  I think that any negative runoff
upstream of these points would send their negative runoff to these points
in the 'direct_to_outlet' case, and then to the ocean.  Positive runoff
would be routed normally, but whatever ends up at an inland sink point will
be sent to the ocean at that location, not at the coast.  It seems like it
would be possible to have long term negative runoff if we specify lakes to
exist, but the climate doesn't have sufficient precip to maintain them.
Because our lakes cannot dry up, they would continue to evaporate.

**Comment by billsacks** [comment]
Thanks, @swensosc, that helps somewhat. However, the figure in question shows large swaths of area with negative runoff - see below. It seems unexpected to me that all of these grid cells would have net negative runoff - or that all of these points would be "sink" points.

<img width="455" alt="image" src="https://user-images.githubusercontent.com/6266741/186251080-2e9df092-1ee3-4021-9314-1628b75de888.png">

**Comment by swensosc** [comment]
I think I misunderstood the question.  The figure you show is the default,
'direct_in_place' behavior, which splits the negative runoff and sends it
'in place' while the positive runoff is routed normally.  In the
'direct_to_outlet' case, some negative runoff is moved to sink points, and
added to whatever positive runoff was routed to that point.  Then it gets
sent to the ocean at that location, whether it's positive or negative.

**Comment by dlawrenncar** [comment]
So I understand better, can you remind me what happens with the negative
runoff in the 'direct_in_place' option.  Does that mean that it is sent
directly to the nearest ocean grid point.  Apologies.  I know you have
answered this before.

Dave

**Comment by swensosc** [comment]
it is sent 'in place', i.e. it is sent at its location, not moved to an
ocean point.  The ocean model (or maybe in the coupler?) has its own
algorithm for finding the nearest ocean point and applying a mask that
spreads out the input over some area (to avoid numerical issues with
dropping it all in one ocean gridcell).

**Comment by dlawrenncar** [comment]
Got it.  That's what I thought.  So, the reality is that in CESM2, when we
move to MOSART, there is a lot more negative runoff being sent to the ocean
than in CESM1 when we sent it to the rivers (RTM, which could handle
negative runoff).  The direct_to_outlet is resolving a lot of that negative
runoff that occurs in CESM2.

**Comment by olyson** [comment]
I'm curious how mosart is able to split negative from positive runoff.  Does it get passed the gridcell average runoff but in component form (surface, drainage, qrgwl, etc.) and then checks for negative runoff from each component at each time step?

**Comment by swensosc** [comment]
read the manual, keith.
seriously though, yes, it checks the components separately each time step
and can do different operations for each.  For example, in the default case
with the 'threshold' option, it will first try to offset negative runoff
from the main channel storage, and if that is not sufficient, it will just
pass the remainder to the coupler.  With the 'negative' option, it doesn't
do that, it just sends the entire negative amount to the coupler.  In the
'direct_to_outlet' case it sends all the negative to the outlet.

**Comment by olyson** [comment]
Ha, yes, I just read the manual.  Point taken :)

**Comment by billsacks** [comment]
Thanks for all this info. I'm still having trouble discerning the short answer to the question. I think what you're saying supports my original feeling that this figure does NOT imply net negative runoff from all of these grid cells, but I'm not sure....

**Comment by olyson** [comment]
MOSART PR issued (https://github.com/ESCOMP/MOSART/pull/57) per @ekluzek request.

**Comment by olyson** [comment]
Regarding the issue I brought up with frozen runoff in the coupler log files (as shown in the last slide of the negative runoff ppt), I ran a clone of the fully coupled direct_to_outlet simulation with Sean's bug fix and that problem appears to be fixed as well.  The frozen runoff from the lnd is showing up in the ocn column rather than the rof column as is normal.

**Comment by wwieder** [comment]
thanks for checking this, Keith.

I still wonder if we addressed @billsacks question on negative net runoff fluxes from amazonian grid cells that we started discussing at co-chairs?

**Comment by olyson** [comment]
I tested the wetland fix in a land-only combined with direct_to_outlet.  Seems to work, there are no longer any wetland areas.  Negative runoff is only reduced by another 1% or so, there isn't that much negative runoff associated with wetlands globally.  The problematic islands don't seem to have negative runoff anymore, at least in a monthly sense.

My understanding now based on our conversation is that the above figure does NOT imply net negative runoff from all of these grid cells.

**Comment by wwieder** [comment]
Thanks for working on this Keith.  
- Are there appreciable changes in energy or water fluxes around our new beachy islands (or coasts)?  
- Are there any relevant updates for co-chairs, beyond "we fixed another tiny fraction of negative runoff"
- Do we need to think about anything else before discussing at the CLM meeting next Thursday?

**Comment by wwieder** [comment]
@billsacks and @ekluzek is this a good issue that should be converted to a discussion?

**Comment by billsacks** [comment]
> is this a good issue that should be converted to a discussion?

Sure.

**Comment by ekluzek** [comment]
@wwieder I think it should. I just wanted to check before hand. I'll go ahead and do that.

**Comment by olyson** [comment]
1. The grid cells that were changed from wetland and small in size (landfrac) and scattered, so I don't think they will affect the simulation much.  I see increases in sensible heat flux and decreases in latent heat flux for those grid cells, which is what we would expect.
2. I don't know of any other updates for co-chairs.
3. If anyone has any other ideas, let me know.

**Comment by olyson** [comment]
Irrigation increases negative runoff by about 15% (compared 10 year I2000 simulations with irrigation on and off).

**Comment by dlawrenncar** [comment]
Can you put this in terms of negative runoff vs total runoff for the
various configurations?  That is, what % of total runoff to the ocean is
negative runoff?  Trying to get a better sense of how big a correction
factor would be if that is what we have to do.

**Comment by olyson** [comment]
Here's a table for land-only:

https://docs.google.com/spreadsheets/d/18NhG4U9ftj9gE0LACDk7lTk5DMEB5Nqm/edit?usp=sharing&ouid=104026900087612515425&rtpof=true&sd=true

Negative runoff is 4.52% of total runoff to ocean (with irrigation on).
Negative runoff is 3.82% of total runoff to ocean (with irrigation off).
This is reduced to 1.45% using the direct_to_outlet + wetland->veg methods.

These numbers are a bit higher than what I reported on in the meeting.  I've updated these numbers to reflect the fact that I'm now using the same land/ocean mesh file that is being used in the coupled simulations for better consistency with those simulations.

For comparison, in one of the 1850 fully coupled simulations (46 years), negative runoff is 3.60% of total runoff to ocean which is pretty consistent with land-only, irrigation-off.

**Comment by wwieder** [comment]
Sorry, one more annoying clarification question,

is the last number with or without irrigation (1.45% negative runoff using
the direct_to_outlet + wetland->veg methods)
If it's with irrigation off, what is negative runoff with irrigation on +
direct_to_outlet + wetland->veg methods?
if it's with irrigation on, what is  negative runoff with irrigation off +
direct_to_outlet + wetland->veg methods?

Would this information be helpful to have?

**Comment by olyson** [comment]
Irrig on (I updated the table).
Easy enough to do an irrig off + direct_to_outlet + wetland->veg methods

**Comment by wwieder** [comment]
Offline conversations we discussed adding a scaling factor prior to exporting runoff. Basically calculate a factor such that:
`global sum(positive_runoff) * factor = sum(all_runoff)`, then set all negative values to zero and apply the factor to the remaining positive values?  

Gokhan prefers this implementation, potentially with some temporal smoothing? 

One remaining question:  In the coupled model, where are runoff fluxes INTO the Caspian mapped to?  Should we create a virtual reservior for these fluxes to help with this smoothing?

**Comment by wwieder** [comment]
Three parts to this plan discussed at the SE meeting

1. **Changing wetland to bare ground** @billsacks has ideas to review this and suggested doing this at runtime (not in the surface dataset), see #1878, and #1886
2.  **Runoff direct to outlet**, this would be a mosart namelist change in coupled simulations @ekluzek will think about how to implement this logic in a coupled model and will propose a solution next week
3. **Remove remaining negative runoff** from global river discharge.  This seems like it should happen in the river model (or the coupler). The later would be nice because it only has to be done once, but may be the most challenging.   This is still TBD.
4. **Negative ice runoff from glaciers** @billsacks mentioned he's supporting Kate T in this too

**Reply by ekluzek** [reply]
"4" in the list above is now filed as #2222

**Comment by slevis-lmwg** [comment]
I am browsing over #1835 here, where @wwieder suggested that I may work on the last checkbox (same as item 3 in the last post), i.e. "remove remaining negative runoff ..."  

Relevant questions (some of them tangential):
1) @billsacks should we close issues https://github.com/ESCOMP/CTSM/issues/1878 and #1716 since you completed and merged the work into ctsm5.2?
2) Whom should I meet with for an orientation on getting started with the last checkbox in https://github.com/ESCOMP/CTSM/discussions/1835?
3) Any reason to prioritize #1886 before the last checkbox in https://github.com/ESCOMP/CTSM/discussions/1835? Just making sure; I'm not concerned one way or another.

**Reply by slevis-lmwg** [reply]
Regarding (1) maybe I was missing that there's a new proposal to do #1878 at runtime? And, therefore, the issue remains open.

**Reply by slevis-lmwg** [reply]
Regarding (3) it seems best to me that we complete #1886 now as part of ctsm5.2. Otherwise we will need to regenerate surface datasets, unless I have misunderstood.

**Reply by billsacks** [reply]
Regarding (1) I think we were leaving the issues open until they are resolved on master - so once the ctsm5.2 branch is merged. I have a slight preference for sticking with that, but if you prefer to go ahead and close them now, that's okay with me. The fix for #1878 does make the changes at runtime, though it was coordinated with some mksurfdata changes.

I think we decided to defer #1886 to the future.

**Comment by billsacks** [comment]
From discussion with @slevis-lmwg and @swensosc today:

We are tentatively planning to remove negative runoff in CMEPS by reducing positive runoff grid cells proportionally to conserve water - see https://github.com/ESCOMP/CMEPS/issues/405. The tentative plan is for me to take this on sometime in the next few months.

We also discussed the idea of adding annual dribblers to negative runoff from lakes and wetlands. We tentatively thought we could defer that piece because:
- Even though that would help reduce negative runoff sent to the runoff model, the feeling was that we might not want kludge-upon-kludge, and the above fix in CMEPS should be sufficient to remove all negative runoff
- Once we have dynamic lakes, this annual dribbling won't be necessary

But we are open to feedback on this.

**Reply by wwieder** [reply]
Thanks for offering to help with this, Bill

**Comment by ekluzek** [comment]
@billsacks this makes sense to me. By putting it in CMEPS it's then handled for any land model that's used. So for example would be handled for SLIM this way. I just have a question.

In CMEPS this won't care what type of landunit is being run, it'll just handle any reason that runoff is negative. We expect this over: glacier, wetland, ocean, and lake, but we could only assess that inside the land model. Inside of CMEPS you won't know what landunit it is.  If you get negative runoff elsewhere that could indicate something is wrong inside of CTSM. I'm just wondering if that's a "con" for this approach? Although on the other hand you could check if runoff is negative over other landunits where it's not expected and still have that inside of CTSM, even with the general handling of negative runoff in CMEPS.

Any reaction to that? I'm kind of thinking the check for negative runoff for landunits that this isn't expected might be the thing to do.

**Reply by billsacks** [reply]
I agree that any checks for negative runoff should be in CTSM. Not only are things separated by landunit there, but they are also separated by different runoff variables, so you can check that certain runoff variables are non-negative while allowing others (like qrgwl) be negative.

**Comment by billsacks** [comment]
I talked to Frank Bryan on the shuttle and he gave a few reasons why negative runoff is problematic. The context was that I said we were wondering how hard we need to work to remove all traces of negative runoff vs just removing most negative runoff. He gave the following reasons why negative runoff is or may be problematic:
- The EBM solves an equation that has a single root if you can assume that runoff is positive; they need to bypass this if they find negative runoff
- Something about modulating nutrient runoff (I think) by scaling based on runoff; it's unclear what to do in the case of negative runoff
- And more generally, with the complexity of MOM, they are concerned that there may be parts of MOM that implicitly assume positive runoff and so may not work correctly in the presence of negative runoff

For all these reasons, he would really like to see no instances of negative runoff so they can just assume that it is always non-negative. In particular, he feels like the solution we have proposed to distribute negative runoff globally (decreasing positive runoff proportionally) would be significantly better than having negative runoff. (Though I realize that there is still a theoretical possibility that we'd end up with a negative runoff residual that cannot be compensated for by positive runoff in this time step, so they may still need conditionals to deal with this unlikely possibility.)

**Comment by wwieder** [comment]
Following up with Gustavo today, he's less worried about the non-physical nature of moving water globally, as long as it gets rid of our negative runoff issues.  @billsacks I'll let you circle back with a plan on handling this in cmeps vs. the river model after conversations with @slevis-lmwg and @mvertens.

**Comment by wwieder** [comment]
I'm closing this discussion, as we believe the issue on the land side has been resolved.
